api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -543,6 +563,122 @@ public int hashCode() {
     }
   }
 
+  public static class GeometryType extends PrimitiveType {
+
+    private final String crs;
+
+    private GeometryType(String crs) {
+      Preconditions.checkNotNull(crs, "CRS cannot be null");
+      this.crs = crs;
+    }
+
+    public static GeometryType get() {
+      return new GeometryType("");
+    }
+
+    public static GeometryType of(String crs) {
+      return new GeometryType(crs == null ? "" : crs);
+    }
+
+    @Override
+    public TypeID typeId() {
+      return TypeID.GEOMETRY;
+    }
+
+    public String crs() {
+      return crs;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      } else if (!(o instanceof GeometryType)) {
+        return false;
+      }
+
+      GeometryType that = (GeometryType) o;
+      return crs.equals(that.crs);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(GeometryType.class, crs);
+    }
+
+    @Override
+    public String toString() {
+      return String.format("geometry(%s)", crs);
+    }
+  }
+
+  public static class GeographyType extends PrimitiveType {
+
+    private final String crs;
+    private final EdgeInterpolationAlgorithm algorithm;
+
+    private GeographyType(String crs, EdgeInterpolationAlgorithm algorithm) {
+      Preconditions.checkNotNull(crs, "CRS cannot be null");
+      this.crs = crs;
+      this.algorithm = algorithm;
+    }
+
+    public static GeographyType get() {
+      return new GeographyType("", null);
+    }
+
+    public static GeographyType of(String crs) {
```

Comment:
Is there a better name than `of` that is more descriptive? I think that `forCRS` would work here, but it is a bit confusing when the algorithm is added (`forCRS("srid:1234", EdgeInterpolationAlgorithm:ANDOYER)`. Maybe `of` is the best choice.

Reply:
I just realized that we should be careful about leading and trailing spaces in `crs`. Should we trim the `crs` before constructing geospatial type objects or throw an exception when leading/trailing spaces are found in `crs`?

Reply:
I decided to throw exception when crs has leading/trailing spaces.

Reply:
if we have two methods (the other one taking crs/algorithm), why not rename this to forCRS now?

Reply:
Renamed to `forCRS`.